### DISCUSSION TITLE
Is there a crate I can use programatically?

### DISCUSSION THREAD

**Discussion by Milo123459** [GitoxideLabs/gitoxide#183] [Q&A] [answered]
I want to use all of the Gitoxide CLI commands without installing the CLI (weird, I know!) I just want to access the APIs of how Gitoxide works, so for example a crate so I can do `gitoxide::pull(..)` etc

**Comment by Byron** [comment, marked as answer]
Thanks for your interest. What you are looking for is the [`gitoxide-core`](https://docs.rs/gitoxide-core/0.9.0/gitoxide_core/), which is what powers the CLI. It allows to control input and output as well, so it can be used with graphical user interfaces as well.

You probably want to build the latest docs from source for now to get an overview, as it currently doesn't build on docs.rs.

**Reply by Milo123459** [reply]
Thank you! Really looking to speed up my project [Glitter](https://github.com/Milo123459/glitter) and this should do it!

**Reply by Byron** [reply]
That's a great project! `gitoxide` would love to power it, but be warned, it will be a while until it's able to especially with things like `pull()`. My recommendation is to follow `gitoxide` and wait for crates of interest to reach [stability tier 2](https://github.com/Byron/gitoxide/blob/c2bc1a6d2b2a1b0ab4963d7edf1b8ab62ba97e4e/STABILITY.md#tiers) or higher at least.

**Reply by Milo123459** [reply]
Ahah, thanks a lot for the kind words! I looked into it, and yes, I completely agree. I'll wait till stability tier 2.
Thanks!

**Reply by Milo123459** [reply]
Quick question, will the `gitoxide::pull` eventually be faster then git pull? If so, that'd be amazing.

**Reply by Byron** [reply]
For clones it already might be as it is able to resolve the downloaded pack file more quickly. For most fetches, this won't play a role though. Lastly, when it will one day checkout files to the work tree, it probably will do it more quickly as it will do so using multiple threads. However, the latest git version already supports multi-threaded checkouts so there might not be much to win.

Note that in order to get that performance, you will have to compile with flags similar to these: 

https://github.com/Byron/gitoxide/blob/262c1229d6d2d55c70fe0e199ab15d10954d967b/Cargo.toml#L44

**Reply by Milo123459** [reply]
I just need the essential features with *speed*. I think what you've suggested shall work!

**Reply by Milo123459** [reply]
Hi, would you mind making a stability issue I can subscribe to so I know when gitoxide reaches stability level 2?

**Reply by Byron** [reply]
Maybe you could watch releases and wait until `git-repository` is 1.0 - GitHub will send you notifications, and with a mail filter you could probably remove all the crates you are not interested in.

**Reply by Milo123459** [reply]
Cool! Can't wait for it :)
